src/native/http_request_response.c

```
@@ -0,0 +1,481 @@
+    }
+
+    (*env)->CallVoidMethod(
+        env,
+        callback->java_crt_http_callback_handler,
+        s_crt_http_stream_handler.onResponseHeaders,
+        callback->java_http_stream,
+        resp_status,
+        jHeaders);
+
+    if ((*env)->ExceptionCheck(env)) {
+        // Close the Connection if the Java Callback throws an Exception
+        aws_http_connection_close(aws_http_stream_get_connection(stream));
+        return;
+    }
+}
+
+static void s_on_incoming_header_block_done_fn(struct aws_http_stream *stream, bool has_body, void *user_data) {
+    struct http_stream_callback_data *callback = (struct http_stream_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback->connection->jvm);
+
+    jboolean jHasBody = has_body;
+    (*env)->CallVoidMethod(
+        env,
+        callback->java_crt_http_callback_handler,
+        s_crt_http_stream_handler.onResponseHeadersDone,
+        callback->java_http_stream,
+        jHasBody);
+
+    if ((*env)->ExceptionCheck(env)) {
+        // Close the Connection if the Java Callback throws an Exception
+        aws_http_connection_close(aws_http_stream_get_connection(stream));
+        return;
+    }
+}
+
+static void s_on_incoming_body_fn(
+    struct aws_http_stream *stream,
+    const struct aws_byte_cursor *data,
+    /* NOLINTNEXTLINE(readability-non-const-parameter) */
+    size_t *out_window_update_size,
+    void *user_data) {
+
+    struct http_stream_callback_data *callback = (struct http_stream_callback_data *)user_data;
+    JNIEnv *env = aws_jni_get_thread_env(callback->connection->jvm);
+
+    jobject jByteBuffer = aws_jni_byte_buffer_copy_from_cursor(env, data);
+    jint window_increment = (*env)->CallIntMethod(
+        env,
+        callback->java_crt_http_callback_handler,
+        s_crt_http_stream_handler.onResponseBody,
+        callback->java_http_stream,
+        jByteBuffer);
+
+    // Verify Window Increment is within bounds
+    assert(0 <= window_increment);
```

These should probably throw an exception or clamp. We throw an exception in .NET.

---

can get rid of the asserts then, you'll crash in debug before delivering the exception, or you'll get a native exception (depending on platform).

---

Fixed